### ISSUE TITLE
Implementing Warm-up and State Reset in Iterative Training

### ISSUE THREAD

**Issue by chooron** [LuxDL/Lux.jl#1312]
In #1284, I have already built an ExpHydro model:

```
using Lux, Random

# smooting step function
step_fct(x) = (tanh.(5.0f0 .* x) .+ 1.0f0) .* 0.5f0
# snow precipitation
Ps(P, T, Tmin) = step_fct(Tmin .- T) .* P
# rain precipitation
Pr(P, T, Tmin) = step_fct(T .- Tmin) .* P
# snow melt
# M(S0, T, Df, Tmax) = min.(S0, Df .* step_fct(T .- Tmax) .* (T .- Tmax))
M(S0, T, Df, Tmax) = Df .* step_fct(T .- Tmax) .* (T .- Tmax)
# evapotranspiration
function PET(T, Lday)
    @. 29.8f0 * Lday * 0.611f0 * exp((17.3f0 * T) / (T + 237.3f0)) / (T + 273.2f0)
end

function ET(S1, T, Lday, Smax)
    return PET(T, Lday) .* S1 ./ Smax .* step_fct(S1 .- Smax) 
end

# base flow
function Qb(S1, f, Smax, Qmax)
    return max.(S1, Qmax .* exp.(-f .* step_fct(Smax .- S1) .* (Smax .- S1)))
end

# peak flow
Qs(S1, Smax) = (S1 .- Smax) .* step_fct(S1 .- Smax)

@kwdef struct ExpHydroLuxLayer <: Lux.AbstractRecurrentCell
    hidden_dims::Int
end

function (l::ExpHydroLuxLayer)(input::AbstractArray, ps, st::NamedTuple)
    return LuxCore.apply(l, (input, st), ps, st)
end

function (l::ExpHydroLuxLayer)(input::Tuple, ps, st::NamedTuple)
    forcing, states = input
    p_vec, t_vec, lday_vec = forcing[1:1, :], forcing[2:2, :], forcing[3:3, :]
    s0_vec, s1_vec = states[1], states[2]
    qout_vec = Qb(s1_vec, ps.f, ps.Smax, ps.Qmax) .+ Qs(s1_vec, ps.Smax)
    melt_vec = M(s0_vec, t_vec, ps.Df, ps.Tmax)
    new_s0_vec = s0_vec .+ Ps(p_vec, t_vec, ps.Tmin) .- melt_vec
    new_s1_vec = s1_vec .+ Pr(p_vec, t_vec, ps.Tmin) .+ melt_vec .- ET(s1_vec, t_vec, lday_vec, ps.Tmax) .- qout_vec
    return (qout_vec, (new_s0_vec, new_s1_vec)), (s0=new_s0_vec, s1=new_s1_vec)
end

function LuxCore.initialparameters(rng::AbstractRNG, l::ExpHydroLuxLayer)
    return (
        f=rand(rng, Float32, 1, l.hidden_dims) .* 0.1f0,
        Smax=rand(rng, Float32, 1, l.hidden_dims) .* (2000.0f0 .- 100.0f0) .+ 100.0f0,
        Qmax=rand(rng, Float32, 1, l.hidden_dims) .* (50.0f0 .- 10.0f0) .+ 10.0f0,
        Df=rand(rng, Float32, 1, l.hidden_dims) .* 5.0f0,
        Tmax=rand(rng, Float32, 1, l.hidden_dims) .* 3.0f0,
        Tmin=rand(rng, Float32, 1, l.hidden_dims) .* -3.0f0,
    )
end

function LuxCore.initialstates(rng::AbstractRNG, l::ExpHydroLuxLayer)
    return (
        s0=rand(rng, Float32, 1, l.hidden_dims),
        s1=rand(rng, Float32, 1, l.hidden_dims),
    )
end

function ExpHydroLuxModel(hidden_dims::Int)
    return Chain(
        Recurrence(
            ExpHydroLuxLayer(hidden_dims); return_sequence=true, ordering=TimeLastIndex()
        ),
        x -> stack(x; dims=3)[1, :, :],
        Dense(hidden_dims, 1, leakyrelu),
    )
end

export ExpHydroLuxModel, ExpHydroLuxLayer
```
The difference is that I need to store the state variables generated by ExpHydroLuxLayer for the computation of the next batch. However, I currently have a new computational requirement: I need to perform a warm-up first during training to obtain stable states, and then reset the states to their initial state after each training iteration. How can this be implemented?
```
x_samples = fill(rand(3,10,365), 30)
y_samples = fill(rand(1,365), 30)
train_state = Training.TrainState(model, ps, init_st, Adam(0.01f0))
for iter in 1:1000
    total_loss, total_sample = 0.0f0, 0
    _, new_st = model(x_samples[1], ps, st)
    @set! train_state.states = new_st
    for (x, y) in zip(x_samples[2:end], y_samples[2:end])
        _, loss, _, train_state = Lux.Training.single_train_step!(
            AutoZygote(), lossfunc,
            (x, y), train_state
        )
        total_loss += loss * length(y)
        total_sample += length(y)
    end
    if iter % 10 == 1 || iter == 100
        @printf "Iteration: %04d \t Loss: %10.9g\n" iter total_loss / total_sample
    end
    @set! train_state.states = init_st
end
```

**Comment by avik-pal** [comment]
Does your current version not work?

**Comment by chooron** [comment]
I'm not sure whether my implementation is correct. I have a few concerns:

In the previous code, the odestates were initialized in a specific way. However, if I need to train the model in batches, I might have to store the final state from the last time step of the previous batch. To do this, I saved that state into the Lux model's states. I'm not sure if this approach is correct.

The reason for batch-wise training is that when using Reactant's compile function, if the time steps in the data exceed 10,000, GPU compilation takes an excessive amount of time. So I'm considering splitting the long time series into chunks of 365 time steps.

During training, my loss becomes NaN. Does this mean the gradient is vanishing? Is this caused by the data or by the code?
